registry/storage/catalog.go

```diff
@@ -3,45 +3,51 @@ package storage
 import (
 	"context"
 	"errors"
+	"fmt"
 	"io"
 	"path"
+	"sort"
 	"strings"
 
 	"github.com/distribution/distribution/v3/reference"
 	"github.com/distribution/distribution/v3/registry/storage/driver"
 )
 
-// Returns a list, or partial list, of repositories in the registry.
+var (
+	// ErrStopReposWalk is a sentinel error to indicate that the repository path walk was stopped.
+	ErrStopReposWalk = errors.New("stop repos walk")
+)
+
+// Repositories populates the passed passed repos slice with repositories in the
+// registry up to the capacity of the slice. Returns the number of repos returned and
+// `io.EOF` if no more repositories are available.
 // Because it's a quite expensive operation, it should only be used when building up
 // an initial set of repositories.
 func (reg *registry) Repositories(ctx context.Context, repos []string, last string) (n int, err error) {
 	var finishedWalk bool
 	var foundRepos []string
 
-	if len(repos) == 0 {
-		return 0, errors.New("no space in slice")
+	if len(repos) == 0 && cap(repos) == 0 {
```

Comment:
Is there a reason why we're checking for slice capacity here? 🤔

Reply:
https://github.com/distribution/distribution/pull/3902#discussion_r1242771306